stl/inc/xutility

```diff
@@ -4889,23 +4889,120 @@ public:
+    _NODISCARD static __forceinline uint64_t __rng_umul128(
+        const uint64_t __a, const uint64_t __b, uint64_t* const __productHi) {
+        // TRANSITION, VSO-634761
+        // The casts here help MSVC to avoid calls to the __allmul library function.
+        const uint32_t __aLo = static_cast<uint32_t>(__a);
+        const uint32_t __aHi = static_cast<uint32_t>(__a >> 32);
+        const uint32_t __bLo = static_cast<uint32_t>(__b);
+        const uint32_t __bHi = static_cast<uint32_t>(__b >> 32);
+
+        const uint64_t __b00 = static_cast<uint64_t>(__aLo) * __bLo;
+        const uint64_t __b01 = static_cast<uint64_t>(__aLo) * __bHi;
+        const uint64_t __b10 = static_cast<uint64_t>(__aHi) * __bLo;
+        const uint64_t __b11 = static_cast<uint64_t>(__aHi) * __bHi;
+
+        const uint32_t __b00Lo = static_cast<uint32_t>(__b00);
+        const uint32_t __b00Hi = static_cast<uint32_t>(__b00 >> 32);
+
+        const uint64_t __mid1   = __b10 + __b00Hi;
+        const uint32_t __mid1Lo = static_cast<uint32_t>(__mid1);
+        const uint32_t __mid1Hi = static_cast<uint32_t>(__mid1 >> 32);
+
+        const uint64_t __mid2   = __b01 + __mid1Lo;
+        const uint32_t __mid2Lo = static_cast<uint32_t>(__mid2);
+        const uint32_t __mid2Hi = static_cast<uint32_t>(__mid2 >> 32);
+
+        const uint64_t __pHi = __b11 + __mid1Hi + __mid2Hi;
+        const uint64_t __pLo = (static_cast<uint64_t>(__mid2Lo) << 32) | __b00Lo;
+
+        *__productHi = __pHi;
+        return __pLo;
+    }
+
+    _NODISCARD static inline uint64_t __rng_shiftright128(
+        const uint64_t __lo, const uint64_t __hi, const uint32_t __dist) {
+        // We don't need to handle the case __dist >= 64 here (see above).
+        _STL_INTERNAL_CHECK(__dist < 64);
+#ifdef _WIN64
+        _STL_INTERNAL_CHECK(__dist > 0);
+        return (__hi << (64 - __dist)) | (__lo >> __dist);
+#else // ^^^ 64-bit ^^^ / vvv 32-bit vvv
+      // Avoid a 64-bit shift by taking advantage of the range of shift values.
+        _STL_INTERNAL_CHECK(__dist >= 32);
+        return (__hi << (64 - __dist)) | (static_cast<uint32_t>(__lo >> 32) >> (__dist - 32));
+#endif // ^^^ 32-bit ^^^
+    }
+
+#endif // ^^^ intrinsics unavailable ^^^
+
+    // using multiply-shift instead of modulo to fairly map values in range [0, _Mod), adapted from arXiv:1805.10941v4
+    static constexpr _Udiff _Reduce(_Udiff _Value, _Udiff _Mod, size_t _Bitrange) noexcept {
+        constexpr auto _Udiff_bits = CHAR_BIT * sizeof(_Udiff);
+        constexpr auto _Uint_bits  = CHAR_BIT * sizeof(unsigned int);
+
+        if constexpr (_Udiff_bits < _Uint_bits) {
+            // avoid unnecessary promotion for smaller types
+            return static_cast<_Udiff>((_Value * _Mod) >> _Bitrange);
+        } else if constexpr (_Udiff_bits == _Uint_bits) {
+            // ensure promotion to ull to keep all bits
+            return static_cast<_Udiff>((static_cast<uint64_t>(_Value) * _Mod) >> _Bitrange);
+        }

```

Comment:
This should be followed by `else`. For ordinary control flow, we prefer to avoid chaining after unconditional `return`, which reduces the amount of nesting. But for `if constexpr`, the `else` prevents that codepath from being emitted at all when the previous condition(s) are selected.

Reply:
OK, will be changed